sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala

```diff
@@ -0,0 +1,442 @@
+    rightLong match {
+      case rightVector: ColumnVector =>
+        checkMultiplyOverflowWithRightCv(leftLong, rightVector, result)
+      case _ =>
+        checkMultiplyOverflowWithRightCv(rightLong, leftLong.asInstanceOf[ColumnVector], result)
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   * if(x == Long.MIN_VALUE && y == -1) throw exception
+   *
+   * @param x long cv or scalar
+   * @param y long cv
+   * @param r is the result of x * y
+   */
+  def checkMultiplyOverflowWithRightCv(x: BinaryOperable, y: ColumnVector,
+      r: ColumnVector): Unit = {
+    // check (y != 0) && (r/y != x)
+    checkMultiplyOverflowSimple(x, y, r)
+
+    // check (x == Long.MIN_VALUE && y == -1)
+    x match {
+      case scalar: Scalar =>
+        if (getLong(scalar) == Long.MinValue) { // x == Long.MIN_VALUE
+          withResource(Scalar.fromLong(-1L)) { negOne =>
+            withResource(y.equalTo(negOne)) { negOneBool =>
+              if (BoolUtils.isAnyValidTrue(negOneBool)) { // y = -1L
+                throw new ArithmeticException("overflow occurs")
+              }
+            }
+          }
+        }
+      case _ =>
+        // check (x == Long.MIN_VALUE && y == -1)
+        withResource(Scalar.fromLong(Long.MinValue)) { minLong =>
+          withResource(Scalar.fromLong(-1L)) { negOneLong =>
+            withResource(x.equalTo(minLong)) { minBool =>
+              withResource(y.equalTo(negOneLong)) { oneBool =>
+                withResource(minBool.and(oneBool)) { overflow =>
+                  if (BoolUtils.isAnyValidTrue(overflow)) {
+                    throw new ArithmeticException("overflow occurs")
+                  }
+                }
+              }
+            }
+          }
+        }
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
+   * r = x * y;
+   * if (y != 0) && (r/y != x) throw exception;
+   *
+   * Here is a little different with `Math.multiplyExact`,
+   * omits the `check x == Long.MIN_VALUE && y == -1`,
+   * Because of the `toIntWithCheck` will throw exception
+   * for the value of `Long.MIN_VALUE * -1` exceeds int limits
+   *
+   * @param leftLong  long cv or scalar
+   * @param rightLong long cv or scalar, will not be scalar if leftLong is scalar
+   * @param result    is the result of x * y
+   */
+  def checkMultiplyOverflowSimple(leftLong: BinaryOperable, rightLong: BinaryOperable,
+      result: ColumnVector): Unit = {
+    rightLong match {
+      case rightVector: ColumnVector =>
+        checkMultiplyOverflowSimpleWithRightCv(leftLong, rightVector, result)
+      case _ =>
+        checkMultiplyOverflowSimpleWithRightCv(rightLong, leftLong.asInstanceOf[ColumnVector],
+          result)
+    }
+  }
+
+  /**
+   * Check if the long result of long * long is overflow
+   * See Math.multiplyExact:
```

Comment:
The overflow checks look pretty expensive to evaluate.  Would it be faster to perform a multiply with a cast to DECIMAL128 (only on the output, if possible) and then check if the upper 64-bits do not match the expected sign bit of the result (i.e.: sign-bit of left XOR sign-bit of right)?  This is something that could be investigated as a followup.

Reply:
Good idea.

I tested the new implementation:
`decimal128_ret = left_long.mul(right_long, DECIMAL128)`  
then check `decimal128_ret < min_long or decimal128_ret > max_long`
then `decimal128_ret.toLong`

|original implementation|new implementation|
|:-:|:-:|
|40ms|25ms|

`DECIMAL_128` fits the results of `Long.MaxValue * Long.MaxValue` and  `Long.MinValue * Long.MinValue`,
and the overflow check `decimal128_ret < min_long or decimal128_ret > max_long` is equavalent to the original implementation.
I think this implementation has no bug, and let's use this.


Details code:
```
  def genLongs(): Array[Long] = {
    val rows = 100 * 1024 * 1024
    val ret = new Array[Long](rows)
    var i = 0
    while (i < rows) {
      ret(i) = Long.MaxValue - i
      i += 1
    }
    ret
  }

  test("multiple decimal 128") {
    val l = genLongs
    val r = genLongs
    val lcv = ColumnVector.fromLongs(l: _*)
    val rcv = ColumnVector.fromLongs(r: _*)

    val start = System.currentTimeMillis()
    val dT = DType.create(DType.DTypeEnum.DECIMAL128, 0)
    withResource(lcv.mul(rcv, dT)) { ret =>
      withResource(Scalar.fromLong(Long.MaxValue)) { maxLong =>
        withResource(Scalar.fromLong(Long.MinValue)) { minLong =>
          withResource(ret.greaterThan(maxLong)) { gtm =>
            withResource(ret.lessThan(minLong)) { ltm =>
              if (BoolUtils.isAnyValidTrue(gtm) || BoolUtils.isAnyValidTrue(ltm)) {
                println("overflow occurs")
              }
              ret.castTo(DType.INT64)
            }
          }
        }
      }
    }
    println((System.currentTimeMillis() - start) + " ms")
  }

  test("multiple long") {
    val l = genLongs
    val r = genLongs
    val lcv = ColumnVector.fromLongs(l: _*)
    val rcv = ColumnVector.fromLongs(r: _*)

    val start = System.currentTimeMillis()
    withResource(lcv.mul(rcv)) { r =>
      // only print if overflow
      IntervalUtils.checkMultiplyOverflow(lcv, rcv, r)
      r
    }

    println((System.currentTimeMillis() - start) + " ms")
  }
```